src/rgw/rgw_auth_s3.cc

```diff
@@ -197,16 +197,15 @@ bool rgw_create_s3_canonical_header(const req_info& info,
   if (qsr) {
     date = info.args.get("Expires");
   } else {
-    const char *str = info.env->get("HTTP_DATE");
+    const char *str = info.env->get("HTTP_X_AMZ_DATE");
     const char *req_date = str;
-    if (str) {
-      date = str;
-    } else {
-      req_date = info.env->get("HTTP_X_AMZ_DATE");
+    if (str == NULL) {
+      req_date = info.env->get("HTTP_DATE");
       if (!req_date) {
         dout(0) << "NOTICE: missing date for auth header" << dendl;
         return false;
       }
+      date = req_date;
```

Comment:
@joke-lee when `str != NULL`, the `date` variable is not assigned so an empty string is passed in the call to `rgw_create_s3_canonical_header()`. this behavior corresponds to the following text from that document:
> When an `x-amz-date` header is present in a request, the system will ignore any `Date` header when computing the request signature. Therefore, if you include the `x-amz-date` header, use the empty string for the `Date` when constructing the `StringToSign`.

but i'm not sure how to reconcile that with the example you included, where the value of the `x-amz-date` header is being used (instead of an empty string) as the `Date` in `StringToSign`